# Стажировка





# Задача: Исследовать датасет

*Для этого нужно подключить библиотеку Numpy.



In [ ]:
# Загрузка модуля pandas
import pandas as pd

Затем нужно создать датасет. Для этого нужно загрузить датасет в Colab.

In [ ]:
# Загрузка файла из облака в colab
import gdown
gdown.download('https://storage.yandexcloud.net/terratraineeship/23_InspectrumClinic_RS/datasets/%D0%B4%D0%B0%D1%82%D0%B0%D1%81%D0%B5%D1%82%20%D0%BF%D1%80%D0%BE%D1%84%D0%BF%D0%B0%D1%82%D0%BE%D0%BB%D0%BE%D0%B3%20%D1%87%D0%B0%D1%81%D1%82%D1%8C.xlsx', None, quiet=True)

Создадаем переменную data, в которую с помощью функции pd.read_excel() загрузим данные из файла profpatolog.xlsx:

**Переименуем**

In [ ]:
import os

# Предположим, что файл загружен с именем "old_filename.txt"
old_filename = "%D0%B4%D0%B0%D1%82%D0%B0%D1%81%D0%B5%D1%82%20%D0%BF%D1%80%D0%BE%D1%84%D0%BF%D0%B0%D1%82%D0%BE%D0%BB%D0%BE%D0%B3%20%D1%87%D0%B0%D1%81%D1%82%D1%8C.xlsx"

# Получение нового имени файла
new_filename = "profpatolog.xlsx"

# Переименование файла
os.rename(old_filename, new_filename)


In [ ]:
data = pd.read_excel('profpatolog.xlsx')


Смотрим что из себя представляет датасет.

In [ ]:
data

ДокументПрохождениеМедосмотра  \
0     7f41ca7e-7dac-11ec-80fc-0cc47aab8067   
1     2ee1f9ac-83dd-11ec-80fc-0cc47aab8067   
2     9cbb191e-8e24-11ec-80fe-0cc47aab8067   
3     b8a33910-8eec-11ec-80fe-0cc47aab8067   
4     0bb0d181-938c-11ec-80fe-0cc47aab8067   
...                                    ...   
1977  137af572-70a5-11ed-8109-0cc47aab8067   
1978  26310d7e-70a5-11ed-8109-0cc47aab8067   
1979  b2602553-705c-11ed-8109-0cc47aab8067   
1980  d0358b83-7065-11ed-8109-0cc47aab8067   
1981  a15e278b-71f5-11ed-810a-0cc47aab8067   

                                    Клиент КлиентДатаРождения КлиентПол  \
0     c91e3407-b81f-11e9-80cf-0cc47aab8067         1976-10-24   Женский   
1     1a1a5fea-8b7f-11eb-80e8-0cc47aab8067         1987-11-18   Мужской   
2     7fe324e0-fa51-11eb-80f6-0cc47aab8067         1985-09-08   Мужской   
3     98f9d761-8d65-11ec-80fe-0cc47aab8067         1988-04-25   Мужской   
4     009aac7d-7bcf-11eb-80e8-0cc47aab8067         1984-04-17   Мужской   
...                                    ...                ...       ...   
1977  41a7c665-7098-11ed-8109-0cc47aab8067         1972-11-12   Мужской   
1978  41a7c663-7098-11ed-8109-0cc47aab8067         1986-07-09   Мужской   
1979  0dc52156-f1a9-11e9-80d1-0cc47aab8067         1994-05-12   Мужской   
1980  6261a674-07c8-11e8-80c6-0cc47aab8067         1972-04-17   Женский   
1981  920293ea-32c2-11eb-80e3-0cc47aab8067         1971-09-14   Мужской   

     КлиентПолПредставление ДатаЗавершенияМедосмотра  ТипМедосмотра  \
0                   Женский               2022-01-25  Периодический   
1                   Мужской               2022-02-02  Периодический   
2                   Мужской               2022-02-15  Периодический   
3                   Мужской               2022-02-16  Периодический   
4                   Мужской               2022-02-22  Периодический   
...                     ...                      ...            ...   
1977                Мужской               2022-11-30  Периодический   
1978                Мужской               2022-11-30  Периодический   
1979                Мужской               2022-11-30  Периодический   
1980                Женский               2022-11-30  Периодический   
1981                Мужской               2022-12-02  Периодический   

     ТипМедосмотраПредставление  ПсихОсвидетельствование  \
0                 Периодический                    False   
1                 Периодический                    False   
2                 Периодический                    False   
3                 Периодический                    False   
4                 Периодический                    False   
...                         ...                      ...   
1977              Периодический                    False   
1978              Периодический                    False   
1979              Периодический                    False   
1980              Периодический                    False   
1981              Периодический                    False   

                                Контрагент  ...  \
0     23c173a7-30d9-11e7-80d8-708bcd7f83cd  ...   
1     a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
2     a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
3     a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
4     a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
...                                    ...  ...   
1977  a9f70e4c-707f-11ed-8109-0cc47aab8067  ...   
1978  a9f70e4c-707f-11ed-8109-0cc47aab8067  ...   
1979  e689a382-17b2-11e9-80ce-0cc47aab8067  ...   
1980  21f2a93d-68f9-11e6-82dc-d897ba0a5c60  ...   
1981  4826f33a-06b9-11ea-80d3-0cc47aab8067  ...   

     Оториноларингология3_Аудиометрия1_Значение  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   
...                                         ...   
1977           

Выведем первые 5 строк из таблицы.

In [ ]:
data.head()

ДокументПрохождениеМедосмотра                                Клиент  \
0  7f41ca7e-7dac-11ec-80fc-0cc47aab8067  c91e3407-b81f-11e9-80cf-0cc47aab8067   
1  2ee1f9ac-83dd-11ec-80fc-0cc47aab8067  1a1a5fea-8b7f-11eb-80e8-0cc47aab8067   
2  9cbb191e-8e24-11ec-80fe-0cc47aab8067  7fe324e0-fa51-11eb-80f6-0cc47aab8067   
3  b8a33910-8eec-11ec-80fe-0cc47aab8067  98f9d761-8d65-11ec-80fe-0cc47aab8067   
4  0bb0d181-938c-11ec-80fe-0cc47aab8067  009aac7d-7bcf-11eb-80e8-0cc47aab8067   

  КлиентДатаРождения КлиентПол КлиентПолПредставление  \
0         1976-10-24   Женский                Женский   
1         1987-11-18   Мужской                Мужской   
2         1985-09-08   Мужской                Мужской   
3         1988-04-25   Мужской                Мужской   
4         1984-04-17   Мужской                Мужской   

  ДатаЗавершенияМедосмотра  ТипМедосмотра ТипМедосмотраПредставление  \
0               2022-01-25  Периодический              Периодический   
1               2022-02-02  Периодический              Периодический   
2               2022-02-15  Периодический              Периодический   
3               2022-02-16  Периодический              Периодический   
4               2022-02-22  Периодический              Периодический   

   ПсихОсвидетельствование                            Контрагент  ...  \
0                    False  23c173a7-30d9-11e7-80d8-708bcd7f83cd  ...   
1                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
2                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
3                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
4                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   

  Оториноларингология3_Аудиометрия1_Значение  \
0                                        NaN   
1                                        NaN   
2                                        NaN   
3                                        NaN   
4                                        NaN   

  Оториноларингология3_Аудиометрия1_ЗначениеПредставление  \
0                                                NaN        
1                                                NaN        
2                                                NaN        
3                                                NaN        
4                                                NaN        

  Оториноларингология1_ЖалобыДополнение_Значение  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

  Оториноларингология1_ЖалобыДополнение_ЗначениеПредставление  \
0                                                NaN            
1                                                NaN            
2                                                NaN            
3                                                NaN            
4                                                NaN            

  Стоматология2_ВП2_Значение Стоматология2_ВП2_ЗначениеПредставление  \
0                        NaN                                     NaN   
1                        NaN                                     NaN   
2                        NaN                                     NaN   
3                        NaN                                     NaN   
4                        NaN                                     NaN   

  Терапия2_ЗаключениеФОГК_Значение  \
0                              NaN   
1                              NaN   
2                              NaN   
3                              NaN   
4                              NaN   

  Терапия2_ЗаключениеФОГК_ЗначениеПредставление  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                              

Крайние 5 строк.

In [ ]:
data.tail()

ДокументПрохождениеМедосмотра  \
1977  137af572-70a5-11ed-8109-0cc47aab8067   
1978  26310d7e-70a5-11ed-8109-0cc47aab8067   
1979  b2602553-705c-11ed-8109-0cc47aab8067   
1980  d0358b83-7065-11ed-8109-0cc47aab8067   
1981  a15e278b-71f5-11ed-810a-0cc47aab8067   

                                    Клиент КлиентДатаРождения КлиентПол  \
1977  41a7c665-7098-11ed-8109-0cc47aab8067         1972-11-12   Мужской   
1978  41a7c663-7098-11ed-8109-0cc47aab8067         1986-07-09   Мужской   
1979  0dc52156-f1a9-11e9-80d1-0cc47aab8067         1994-05-12   Мужской   
1980  6261a674-07c8-11e8-80c6-0cc47aab8067         1972-04-17   Женский   
1981  920293ea-32c2-11eb-80e3-0cc47aab8067         1971-09-14   Мужской   

     КлиентПолПредставление ДатаЗавершенияМедосмотра  ТипМедосмотра  \
1977                Мужской               2022-11-30  Периодический   
1978                Мужской               2022-11-30  Периодический   
1979                Мужской               2022-11-30  Периодический   
1980                Женский               2022-11-30  Периодический   
1981                Мужской               2022-12-02  Периодический   

     ТипМедосмотраПредставление  ПсихОсвидетельствование  \
1977              Периодический                    False   
1978              Периодический                    False   
1979              Периодический                    False   
1980              Периодический                    False   
1981              Периодический                    False   

                                Контрагент  ...  \
1977  a9f70e4c-707f-11ed-8109-0cc47aab8067  ...   
1978  a9f70e4c-707f-11ed-8109-0cc47aab8067  ...   
1979  e689a382-17b2-11e9-80ce-0cc47aab8067  ...   
1980  21f2a93d-68f9-11e6-82dc-d897ba0a5c60  ...   
1981  4826f33a-06b9-11ea-80d3-0cc47aab8067  ...   

     Оториноларингология3_Аудиометрия1_Значение  \
1977                                        NaN   
1978                                        NaN   
1979                                        NaN   
1980                                        NaN   
1981                                        NaN   

     Оториноларингология3_Аудиометрия1_ЗначениеПредставление  \
1977                                                NaN        
1978                                                NaN        
1979                                                NaN        
1980                                                NaN        
1981                                                NaN        

     Оториноларингология1_ЖалобыДополнение_Значение  \
1977                                            NaN   
1978                                            NaN   
1979                                            NaN   
1980                                            NaN   
1981                                            NaN   

     Оториноларингология1_ЖалобыДополнение_ЗначениеПредставление  \
1977                                                NaN            
1978                                                NaN            
1979                                                NaN            
1980                                                NaN            
1981                                                NaN            

     Стоматология2_ВП2_Значение Стоматология2_ВП2_ЗначениеПредставление  \
1977                        NaN                                     NaN   
1978                        NaN                                     NaN   
1979                        NaN                                     NaN   
1980                        NaN                                     NaN   
1981                        NaN                                     NaN   

     Терапия2_ЗаключениеФОГК_Значение  \
1977                              NaN   
1978                              NaN   
1979                              NaN   
1980                              NaN   
1981                              NaN   

     Терапия2_ЗаключениеФОГК_ЗначениеПредставление  \
1977  

Первично датасет готов к обработке.

# Анализ датасета

Проанализируем данные для нейронной сети.

In [ ]:
data.shape[1]

4587

Имеет 1982 строки и 4587 столбцов .

In [ ]:
# Сколько элементов в таблице
data.size 

9091434

In [ ]:
data.count()

ДокументПрохождениеМедосмотра                                             1982
Клиент                                                                    1982
КлиентДатаРождения                                                        1982
КлиентПол                                                                 1982
КлиентПолПредставление                                                    1982
                                                                          ... 
Стоматология2_ВП2_ЗначениеПредставление                                      1
Терапия2_ЗаключениеФОГК_Значение                                             1
Терапия2_ЗаключениеФОГК_ЗначениеПредставление                                1
Офтальмология1_ДвигательныйАппаратГлазДополнение_Значение                    1
Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление       1
Length: 4587, dtype: int64

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1982 entries, 0 to 1981
Columns: 4587 entries, ДокументПрохождениеМедосмотра to Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление
dtypes: bool(1), datetime64[ns](43), float64(1253), object(3290)
memory usage: 69.3+ MB


DataFrame состоит из 1982 записей (строк) и 4587 столбцов.

Индекс (индексная колонка) имеет тип RangeIndex и содержит значения от 0 до 1981.
В DataFrame присутствуют различные типы данных:
1 cтолбец имеет тип bool.

43 столбца имеют тип datetime64[ns], что указывает на наличие дат и времени.

1253 столбца имеют тип float64, что указывает на наличие числовых значений с плавающей запятой.

3290 столбцов имеют тип object, что обычно означает наличие строковых данных или смешанных типов данных.

Общий объем памяти, используемый для хранения DataFrame, составляет примерно 69,3 мегабайта.

In [ ]:
#Смотрим есть ли пустые ячейки
data.isna().head()

ДокументПрохождениеМедосмотра  Клиент  КлиентДатаРождения  КлиентПол  \
0                          False   False               False      False   
1                          False   False               False      False   
2                          False   False               False      False   
3                          False   False               False      False   
4                          False   False               False      False   

   КлиентПолПредставление  ДатаЗавершенияМедосмотра  ТипМедосмотра  \
0                   False                     False          False   
1                   False                     False          False   
2                   False                     False          False   
3                   False                     False          False   
4                   False                     False          False   

   ТипМедосмотраПредставление  ПсихОсвидетельствование  Контрагент  ...  \
0                       False                    False       False  ...   
1                       False                    False       False  ...   
2                       False                    False       False  ...   
3                       False                    False       False  ...   
4                       False                    False       False  ...   

   Оториноларингология3_Аудиометрия1_Значение  \
0                                        True   
1                                        True   
2                                        True   
3                                        True   
4                                        True   

   Оториноларингология3_Аудиометрия1_ЗначениеПредставление  \
0                                               True         
1                                               True         
2                                               True         
3                                               True         
4                                               True         

   Оториноларингология1_ЖалобыДополнение_Значение  \
0                                            True   
1                                            True   
2                                            True   
3                                            True   
4                                            True   

   Оториноларингология1_ЖалобыДополнение_ЗначениеПредставление  \
0                                               True             
1                                               True             
2                                               True             
3                                               True             
4                                               True             

   Стоматология2_ВП2_Значение  Стоматология2_ВП2_ЗначениеПредставление  \
0                        True                                     True   
1                        True                                     True   
2                        True                                     True   
3                        True                                     True   
4                        True                                     True   

   Терапия2_ЗаключениеФОГК_Значение  \
0                              True   
1                              True   
2                              True   
3                              True   
4                              True   

   Терапия2_ЗаключениеФОГК_ЗначениеПредставление  \
0                                           True   
1                                           True   
2                                           True   
3                                           True   
4                                           True   

   Офтальмология1_ДвигательныйАппаратГлазДополнение_Значение  \
0                                               True           
1                                               True           
2                                               True           
3                                              

In [ ]:
#Сколько пустых ячеек?
data.isna().sum()

ДокументПрохождениеМедосмотра                                                0
Клиент                                                                       0
КлиентДатаРождения                                                           0
КлиентПол                                                                    0
КлиентПолПредставление                                                       0
                                                                          ... 
Стоматология2_ВП2_ЗначениеПредставление                                   1981
Терапия2_ЗаключениеФОГК_Значение                                          1981
Терапия2_ЗаключениеФОГК_ЗначениеПредставление                             1981
Офтальмология1_ДвигательныйАппаратГлазДополнение_Значение                 1981
Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление    1981
Length: 4587, dtype: int64

Вообщем и целом видим какое-то количество пропусков есть.

In [ ]:
#Доля пропусков от общего количества элементов
data.isna().sum() / data.shape[0]

ДокументПрохождениеМедосмотра                                             0.000000
Клиент                                                                    0.000000
КлиентДатаРождения                                                        0.000000
КлиентПол                                                                 0.000000
КлиентПолПредставление                                                    0.000000
                                                                            ...   
Стоматология2_ВП2_ЗначениеПредставление                                   0.999495
Терапия2_ЗаключениеФОГК_Значение                                          0.999495
Терапия2_ЗаключениеФОГК_ЗначениеПредставление                             0.999495
Офтальмология1_ДвигательныйАппаратГлазДополнение_Значение                 0.999495
Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление    0.999495
Length: 4587, dtype: float64

In [ ]:
# Общая сумма пропусков
data.isna().sum().sum()

8106978

In [ ]:
# Общий процент пропусков
proc = data.isna().sum().sum() / data.size
print('{:.1%}'.format(proc))

89.2%


Исходя из этих данных получается что пропущено 90% строк?

Параметр include=['O'] метода .describe() выведет статистику для строковых данных:

*   count - количество заполненных элементов
*   unique - количество уникальных элементов
*   top - самый распространенный элемент
*   freq - сколько раз встречается самый распространенный элемент


In [ ]:
# Вывод статистики для строковых данных.
data.describe(include=['O'])

ДокументПрохождениеМедосмотра  \
count                                   1982   
unique                                  1915   
top     e334cb0e-55a6-11ed-8108-0cc47aab8067   
freq                                       2   

                                      Клиент КлиентПол КлиентПолПредставление  \
count                                   1982      1982                   1982   
unique                                  1911         2                      2   
top     116fe6a1-455f-11ed-8107-0cc47aab8067   Мужской                Мужской   
freq                                       3      1451                   1451   

        ТипМедосмотра ТипМедосмотраПредставление  \
count            1982                       1982   
unique              2                          2   
top     Периодический              Периодический   
freq             1981                       1981   

                                  Контрагент         Цех Профессия  \
count                                   1982         802      1982   
unique                                   177         183       690   
top     173d6ba7-12fe-11e7-80d3-708bcd7f83cd  Без отдела  Водитель   
freq                                     162          63        55   

       ЗаключениеМК  ... Оториноларингология3_Аудиометрия1_Значение  \
count          1575  ...                                          1   
unique            9  ...                                          1   
top           Годен  ...       2934f2d1-8e1f-11ec-80fe-0cc47aab8067   
freq           1434  ...                                          1   

       Оториноларингология3_Аудиометрия1_ЗначениеПредставление  \
count                                                   1        
unique                                                  1        
top                              не подлежит обследованию        
freq                                                    1        

       Оториноларингология1_ЖалобыДополнение_Значение  \
count                                               1   
unique                                              1   
top                                   на ком в горле.   
freq                                                1   

       Оториноларингология1_ЖалобыДополнение_ЗначениеПредставление  \
count                                                   1            
unique                                                  1            
top                                       на ком в горле.            
freq                                                    1            

       Стоматология2_ВП2_Значение Стоматология2_ВП2_ЗначениеПредставление  \
count                           1                                       1   
unique                          1                                       1   
top                             C                                       C   
freq                            1                                       1   

                         Терапия2_ЗаключениеФОГК_Значение  \
count                                                   1   
unique                                                  1   
top     Рекомендации: лекарственные препараты принимат...   
freq                                                    1   

            Терапия2_ЗаключениеФОГК_ЗначениеПредставление  \
count                                                   1   
unique                                                  1   
top     Рекомендации: лекарственные препараты принимат...   
freq                                                    1   

       Офтальмология1_ДвигательныйАппаратГлазДополнение_Значение  \
count                                                   1          
unique                                                  1          
top           девиация левого глаза к носу до 25 градусов          
freq                                                    1          

       Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление  
count        

**Выводы сделанные на основе анализа датасета**

*   Имеет 1982 строки и 4587 столбцов .
*   Общие данные после команды data.inf()
*   Почему-то общий процент пропусков 90%.
*   Есть уникальные идентификаторы. Они не нужны.
*   Есть закономерности, такие как пол влияет на прохождение врачей.
*   Нас иньересуют только 3 класса-**годен,ограниченно годен,негоден**.
*   Если не прошел психОсведетельствование это **Негоден**.


# Создание выборок

In [ ]:
# Создадим копию, так как будем редактировать таблицу.
data_backup = data.copy()
data.head()

ДокументПрохождениеМедосмотра                                Клиент  \
0  7f41ca7e-7dac-11ec-80fc-0cc47aab8067  c91e3407-b81f-11e9-80cf-0cc47aab8067   
1  2ee1f9ac-83dd-11ec-80fc-0cc47aab8067  1a1a5fea-8b7f-11eb-80e8-0cc47aab8067   
2  9cbb191e-8e24-11ec-80fe-0cc47aab8067  7fe324e0-fa51-11eb-80f6-0cc47aab8067   
3  b8a33910-8eec-11ec-80fe-0cc47aab8067  98f9d761-8d65-11ec-80fe-0cc47aab8067   
4  0bb0d181-938c-11ec-80fe-0cc47aab8067  009aac7d-7bcf-11eb-80e8-0cc47aab8067   

  КлиентДатаРождения КлиентПол КлиентПолПредставление  \
0         1976-10-24   Женский                Женский   
1         1987-11-18   Мужской                Мужской   
2         1985-09-08   Мужской                Мужской   
3         1988-04-25   Мужской                Мужской   
4         1984-04-17   Мужской                Мужской   

  ДатаЗавершенияМедосмотра  ТипМедосмотра ТипМедосмотраПредставление  \
0               2022-01-25  Периодический              Периодический   
1               2022-02-02  Периодический              Периодический   
2               2022-02-15  Периодический              Периодический   
3               2022-02-16  Периодический              Периодический   
4               2022-02-22  Периодический              Периодический   

   ПсихОсвидетельствование                            Контрагент  ...  \
0                    False  23c173a7-30d9-11e7-80d8-708bcd7f83cd  ...   
1                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
2                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
3                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   
4                    False  a6d559b0-7835-11eb-80e8-0cc47aab8067  ...   

  Оториноларингология3_Аудиометрия1_Значение  \
0                                        NaN   
1                                        NaN   
2                                        NaN   
3                                        NaN   
4                                        NaN   

  Оториноларингология3_Аудиометрия1_ЗначениеПредставление  \
0                                                NaN        
1                                                NaN        
2                                                NaN        
3                                                NaN        
4                                                NaN        

  Оториноларингология1_ЖалобыДополнение_Значение  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

  Оториноларингология1_ЖалобыДополнение_ЗначениеПредставление  \
0                                                NaN            
1                                                NaN            
2                                                NaN            
3                                                NaN            
4                                                NaN            

  Стоматология2_ВП2_Значение Стоматология2_ВП2_ЗначениеПредставление  \
0                        NaN                                     NaN   
1                        NaN                                     NaN   
2                        NaN                                     NaN   
3                        NaN                                     NaN   
4                        NaN                                     NaN   

  Терапия2_ЗаключениеФОГК_Значение  \
0                              NaN   
1                              NaN   
2                              NaN   
3                              NaN   
4                              NaN   

  Терапия2_ЗаключениеФОГК_ЗначениеПредставление  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                              

In [ ]:
#Восстановление датасета.
data = data_backup.copy()
data.head()

Хочу удалить пустые столбцы.

In [ ]:
import pandas as pd

# Удаление столбцов, в которых все значения являются пустыми
data_without_empty_columns = data.dropna(axis=1, how='all')

# Удаление строк, в которых все значения являются пустыми
data_cleaned = data_without_empty_columns.dropna(axis=0, how='all')

# В результате data_cleaned будет содержать данные без пустых строк и столбцов, в которых все значения являются пустыми


In [ ]:
# Общий процент пропусков
proc = data_cleaned.isna().sum().sum() / data.size
print('{:.1%}'.format(proc))

69.9%


Посмотрим общий процент заполненности строк. Может поможет.

In [ ]:

# Получение общего количества столбцов
total_columns = len(data_cleaned.columns)

# Инициализация переменных для подсчета суммы процентов заполненности
total_filled_percentage = 0
total_rows = 0

# Проход по каждой строке
for index, row in data_cleaned.iterrows():
    # Подсчет количества непустых значений в строке
    filled_values = row.count()
    
    # Подсчет процента заполненности строки
    filled_percentage = (filled_values / total_columns) * 100
    
    # Обновление суммы процентов заполненности и количества строк
    total_filled_percentage += filled_percentage
    total_rows += 1

# Вычисление среднего процента заполненности строк
average_filled_percentage = total_filled_percentage / total_rows

print("Средний процент заполненности строк:", average_filled_percentage)


Средний процент заполненности строк: 13.413402229572085


In [ ]:
data_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1982 entries, 0 to 1981
Columns: 3703 entries, ДокументПрохождениеМедосмотра to Офтальмология1_ДвигательныйАппаратГлазДополнение_ЗначениеПредставление
dtypes: bool(1), datetime64[ns](43), float64(369), object(3290)
memory usage: 56.0+ MB


Уиды из 1С содержат по 36 символов, попробуем удалить их вместе со столбцами.

**Удалю ненужные столбцы.**

In [ ]:
# Указать список столбцов, которые нужно удалить
columns_to_drop = ['ДокументПрохождениеМедосмотра', 'Клиент', 'Контрагент']

# Удалить указанные столбцы из датасета
data_cleaned.drop(columns=columns_to_drop, inplace=True)


In [ ]:
data_cleaned.head()

КлиентДатаРождения КлиентПол КлиентПолПредставление  \
0         1976-10-24   Женский                Женский   
1         1987-11-18   Мужской                Мужской   
2         1985-09-08   Мужской                Мужской   
3         1988-04-25   Мужской                Мужской   
4         1984-04-17   Мужской                Мужской   

  ДатаЗавершенияМедосмотра  ТипМедосмотра ТипМедосмотраПредставление  \
0               2022-01-25  Периодический              Периодический   
1               2022-02-02  Периодический              Периодический   
2               2022-02-15  Периодический              Периодический   
3               2022-02-16  Периодический              Периодический   
4               2022-02-22  Периодический              Периодический   

   ПсихОсвидетельствование                                                Цех  \
0                    False                                                NaN   
1                    False                                                NaN   
2                    False                                                NaN   
3                    False   Отдел техники и технологии строительства скважин   
4                    False  Отдел технологий текущего и капитального ремон...   

                                   Профессия            ЗаключениеМК  ...  \
0  Ведущий специалист по организации обуения                   Годен  ...   
1                         Ведущий специалист                   Годен  ...   
2                           Главный менеджер  ГоденСКоррекциейЗрения  ...   
3                         Главный специалист                   Годен  ...   
4                         Старший специалист  ГоденСКоррекциейЗрения  ...   

  Оториноларингология3_Аудиометрия1_Значение  \
0                                        NaN   
1                                        NaN   
2                                        NaN   
3                                        NaN   
4                                        NaN   

  Оториноларингология3_Аудиометрия1_ЗначениеПредставление  \
0                                                NaN        
1                                                NaN        
2                                                NaN        
3                                                NaN        
4                                                NaN        

  Оториноларингология1_ЖалобыДополнение_Значение  \
0                                            NaN   
1                                            NaN   
2                                            NaN   
3                                            NaN   
4                                            NaN   

  Оториноларингология1_ЖалобыДополнение_ЗначениеПредставление  \
0                                                NaN            
1                                                NaN            
2                                                NaN            
3                                                NaN            
4                                                NaN            

  Стоматология2_ВП2_Значение Стоматология2_ВП2_ЗначениеПредставление  \
0                        NaN                                     NaN   
1                        NaN                                     NaN   
2                        NaN                                     NaN   
3                        NaN                                     NaN   
4                        NaN                                     NaN   

  Терапия2_ЗаключениеФОГК_Значение  \
0                              NaN   
1                              NaN   
2                              NaN   
3                              NaN   
4                              NaN   

  Терапия2_ЗаключениеФОГК_ЗначениеПредставление  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   

Нужно узнать сколько и каких типов данных в столбце ЗаключениеМК, так как эти данные нужно поделить на 3 класса и использовать их затем для обучения нейронки.

In [ ]:
data_cleaned['ЗаключениеМК'].value_counts()


Годен                               1434
ГоденСКоррекциейЗрения               123
ВременноНегоден                        9
НуждаетсяВДообследованииИЛечении       2
ОграниченноГоден                       2
Газоспасатель                          2
ГоденБезРаботНаВысотах                 1
машинист бульдозера                    1
 Газоспасатель                         1
Name: ЗаключениеМК, dtype: int64

Обнаружен машинист бульдозера и газоспасатель! Причем один Газоспасатель неуёмный оказался и не хотел удаляться.


In [ ]:
data_cleaned = data_cleaned[~data_cleaned['ЗаключениеМК'].isin(['машинист бульдозера', 'Газоспасатель', ' Газоспасатель'])]


Нужно оставшееся превести к 3 значениям. Годен, Годен с ограничениями, Негоден.

In [ ]:
data_cleaned['ЗаключениеМК'] = data_cleaned['ЗаключениеМК'].replace({
    'ГоденСКоррекциейЗрения': 'Годен с ограничениями',
    'ОграниченноГоден': 'Годен с ограничениями',
    'ГоденБезРаботНаВысотах': 'Годен с ограничениями',
    'ВременноНегоден': 'Негоден',
    'НуждаетсяВДообследованииИЛечении': 'Негоден'
})



**Получилось!**

Годен                    1434

Годен с ограничениями     126

Негоден                    11

Name: ЗаключениеМК, dtype: int64

Разбалансирован немного, но ничего. Можно пробовать разбить на выборки.

**Попробуем поколдовать при создании выборки для обучения.**

In [ ]:
from sklearn.model_selection import train_test_split

# Разделение данных на признаки (X) и целевую переменную (y)
X = data_cleaned.drop(columns=['ЗаключениеМК'])
y = data_cleaned['ЗаключениеМК']

# Разделение на обучающую выборку, валидационную выборку и тестовую выборку
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_test, y_test, test_size=0.5, random_state=42)


Прводим к одному типу.

In [ ]:
import numpy as np

# Приведение типа данных целевой переменной к строковому типу
y_train = y_train.astype(str)
y_val = y_val.astype(str)
y_test = y_test.astype(str)

print("Уникальные классы в y_train:", len(np.unique(y_train)))
print("Уникальные классы в y_val:", len(np.unique(y_val)))
print("Уникальные классы в y_test:", len(np.unique(y_test)))


Уникальные классы в y_train: 4
Уникальные классы в y_val: 4
Уникальные классы в y_test: 4


Проверяем соответсвие количеству кклассов. Должно быть 3.
Откуда он взялся неясно. Попробуем выяснить названия.

In [ ]:
import numpy as np

# Получение уникальных значений из целевой переменной
unique_classes_train = np.unique(y_train)
unique_classes_val = np.unique(y_val)
unique_classes_test = np.unique(y_test)

# Вывод названий классов
print("Названия классов в y_train:", unique_classes_train)
print("Названия классов в y_val:", unique_classes_val)
print("Названия классов в y_test:", unique_classes_test)


Названия классов в y_train: ['nan' 'Годен' 'Годен с ограничениями' 'Негоден']
Названия классов в y_val: ['nan' 'Годен' 'Годен с ограничениями' 'Негоден']
Названия классов в y_test: ['nan' 'Годен' 'Годен с ограничениями' 'Негоден']


Это был присловутый NAN. Надо от него избавится.

In [ ]:
# Отфильтровать образцы с классом 'nan' из выборок
filter_train = y_train != 'nan'
filter_val = y_val != 'nan'
filter_test = y_test != 'nan'

X_train_filtered = X_train[filter_train]
y_train_filtered = y_train[filter_train]

X_val_filtered = X_val[filter_val]
y_val_filtered = y_val[filter_val]

X_test_filtered = X_test[filter_test]
y_test_filtered = y_test[filter_test]


Проверим сколько их стало этих классов.

In [ ]:
import numpy as np

unique_classes_train = np.unique(y_train_filtered)
unique_classes_val = np.unique(y_val_filtered)
unique_classes_test = np.unique(y_test_filtered)

print("Уникальные классы в y_train_filtered:", len(unique_classes_train))
print("Уникальные классы в y_val_filtered:", len(unique_classes_val))
print("Уникальные классы в y_test_filtered:", len(unique_classes_test))


Уникальные классы в y_train_filtered: 3
Уникальные классы в y_val_filtered: 3
Уникальные классы в y_test_filtered: 3


Отличтно стало 3. Как нам и нужно.

Создали, теперь токенизируем.

**Токенизация**

Прводим к числовому значению.

In [ ]:
from sklearn.preprocessing import LabelEncoder
from keras.utils import to_categorical

# Создание экземпляра LabelEncoder
label_encoder = LabelEncoder()

# Преобразование строковых классов в числовые значения
y_train_encoded = label_encoder.fit_transform(y_train_filtered)
y_val_encoded = label_encoder.transform(y_val_filtered)
y_test_encoded = label_encoder.transform(y_test_filtered)

# Преобразование числовых значений в формат OHE
y_train_encoded = to_categorical(y_train_encoded, num_classes)
y_val_encoded = to_categorical(y_val_encoded, num_classes)
y_test_encoded = to_categorical(y_test_encoded, num_classes)


преобразуем паленые символы еще раз.

In [ ]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from keras.utils import to_categorical

# Создание экземпляра LabelEncoder
label_encoder = LabelEncoder()

# Фильтрация данных
valid_indices = ~y_train_filtered.isna()
y_train_filtered = y_train_filtered[valid_indices].reset_index(drop=True)
X_train_filtered = X_train.loc[valid_indices].reset_index(drop=True)

# Преобразование строковых классов в числовые значения
y_train_encoded = label_encoder.fit_transform(y_train_filtered)
y_val_encoded = label_encoder.transform(y_val_filtered)
y_test_encoded = label_encoder.transform(y_test_filtered)

# Преобразование числовых значений в формат OHE
num_classes = 3  # Замените на фактическое количество классов
y_train_encoded = to_categorical(y_train_encoded, num_classes)
y_val_encoded = to_categorical(y_val_encoded, num_classes)
y_test_encoded = to_categorical(y_test_encoded, num_classes)


IndexingError: ignored

Приведем к OHE.

In [ ]:
from keras.utils import to_categorical

# Определение количества классов
num_classes = len(unique_classes_train)

# Преобразование целевых переменных в формат OHE
y_train_encoded = to_categorical(y_train_filtered, num_classes)
y_val_encoded = to_categorical(y_val_filtered, num_classes)
y_test_encoded = to_categorical(y_test_filtered, num_classes)


ValueError: ignored

не получается. надо узнать в чем дело.

Проверяем как затокенезировалось.

In [ ]:
# Вывод примера токенизированных и закодированных данных
print("Пример токенизированных данных:")
print(X_train_tokenized.iloc[:5])
print("\nПример закодированных данных:")
print(X_train_encoded[:5])


Пример токенизированных данных:
КлиентДатаРождения          [906, 1983-11-04, 777, 1981-12-23, 1623, 1987-...
КлиентПол                   [906, Мужской, 777, Женский, 1623, Мужской, 15...
КлиентПолПредставление      [906, Мужской, 777, Женский, 1623, Мужской, 15...
ДатаЗавершенияМедосмотра    [906, 2022-11-09, 777, 2022-11-08, 1623, 2022-...
ТипМедосмотра               [906, Периодический, 777, Периодический, 1623,...
dtype: object

Пример закодированных данных:
[[0. 0. 0. ... 1. 0. 1.]
 [0. 0. 0. ... 1. 0. 1.]
 [0. 0. 0. ... 1. 0. 1.]
 [0. 0. 0. ... 1. 0. 1.]
 [0. 0. 0. ... 1. 0. 1.]]


Вроде всё закодировалось и токенезировалось.

Проверим еще.

In [ ]:
print("Форма закодированной обучающей выборки:", X_train_encoded.shape)
print("Форма закодированной валидационной выборки:", X_val_encoded.shape)
print("Форма закодированной тестовой выборки:", X_test_encoded.shape)


Форма закодированной обучающей выборки: (1384, 57849)
Форма закодированной валидационной выборки: (297, 57849)
Форма закодированной тестовой выборки: (297, 57849)


Исходя из проверки, вроде какито числа в скобочках:)
А значит всё нормально)

**Нейронка**

Перекодируем названия. Чтобы не писать вручную.

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Инициализация и обучение кодировщика меток классов
label_encoder = LabelEncoder()
label_encoder.fit(y_train)

# Преобразование меток классов в числовые значения
y_train_encoded = label_encoder.transform(y_train)
y_val_encoded = label_encoder.transform(y_val)
y_test_encoded = label_encoder.transform(y_test)


In [ ]:
from keras.utils import to_categorical

# Преобразование целевой переменной в формат one-hot encoding
num_classes = 3  # Замените на фактическое количество классов
y_train_encoded = to_categorical(y_train_encoded, num_classes)
y_val_encoded = to_categorical(y_val_encoded, num_classes)
y_test_encoded = to_categorical(y_test_encoded, num_classes)


IndexError: ignored

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

# Создание модели нейронной сети
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=(X_train_encoded.shape[1],)))
model.add(Dropout(0.2))
model.add(Dense(32, activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(3, activation='softmax'))

# Компиляция модели
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# Обучение модели
history = model.fit(X_train_encoded, y_train_encoded, epochs=10, batch_size=32, validation_data=(X_val_encoded, y_val_encoded))

# Оценка модели на тестовых данных
loss, accuracy = model.evaluate(X_test_encoded, y_test_encoded)
print("Потери (Loss):", loss)
print("Точность (Accuracy):", accuracy)


Epoch 1/10


ValueError: ignored